### PR TITLE
Support other type of webpack configs

### PR DESCRIPTION
**Pull request by Ailrun**
Issue: #3780.

## What I did
Add `interpret` package to support other type (`.ts`, `.babel.js`, ...) of webpack config files.

## How to test
Install `ts-node` and `typescript` in `official-storybook` in example directory, and change `webpack.config.js` to `webpack.config.ts`, and finally change `require`s to `import`s.

Is this testable with Jest or Chromatic screenshots? No
Does this need a new example in the kitchen sink apps? No
Does this need an update to the documentation? I'm not sure.

### CODE REVIEW

**Comment by Ailrun** [comment]
@Hypnosphi It looks like the check fails because of yarn registry. Could you rerun it?

**Review by Hypnosphi — APPROVED** [review]
I'm not very comfortable with copy-pasting the code from webpack sources, but it's probably the best we can do now.

Can you please open an issue for https://github.com/webpack/webpack-cli and ask them to make config resolution logic available for external usage? Babel does a similar thing in v7: https://github.com/babel/babel/pull/7472

**Comment by Ailrun** [comment]
@Hypnosphi the problem is that storybook can accept different type of configurations (even the file names are same)

File: lib/core/src/server/config.js [hunk 1]
```diff
@@ -58,6 +59,87 @@ function informAboutCustomConfig(defaultConfigName) {
   logger.info(`=> Using default webpack setup based on "${defaultConfigName}".`);
 }
 
+// Copied and modified from
```
Comment (igor-dv) [inline lib/core/src/server/config.js]: Please extract all these to a separate file.

**Review by igor-dv — CHANGES_REQUESTED** [review]

**Comment by Hypnosphi** [comment]
Yes, we only need to resolve the filename

File: lib/core/src/server/config.js (hunk 1 above)
Comment (Ailrun) [reply lib/core/src/server/config.js]: OK, I will extract them into `webpack_config`. Is it what you expect?

**Comment by Ailrun** [comment]
@igor-dv Could you check this again?

**Review by igor-dv — APPROVED** [review]
Let's rename `webpack_config` to the `loadCustomWebpackConfig` and we are OK.

**Comment by Ailrun** [comment]
@igor-dv Is there any reason for snake_case file name and camelCase file name? I mean, `babel_config` use snake_case, where `loadCustomWebpackConfig` use camelCase.

**Comment by igor-dv** [comment]
I am not surer about the convention, but it's more about maintenance/readability. I would also change the `babel_config` to `loadBabelConfig` because it confuses.

**Comment by nickmccurdy** [comment]
I think I found a regression:
```
info @storybook/react v4.0.0-alpha.21
info
info => Loading presets
WARN => File /Users/nick/Repos/ui/.storybook/webpack.config.ts is detected
WARN    but impossible to import loader for .ts

```

**Comment by 87Hz** [comment]
> I think I found a regression:
> 
> ```
> info @storybook/react v4.0.0-alpha.21
> info
> info => Loading presets
> WARN => File /Users/nick/Repos/ui/.storybook/webpack.config.ts is detected
> WARN    but impossible to import loader for .ts
> ```

Do not forget to install `@babel/register`.

**Comment by kristojorg** [comment]
@87Hz What do you mean? Why am I installing that and where would I use it?

**Comment by 87Hz** [comment]
> @87Hz What do you mean? Why am I installing that and where would I use it?
> 
> I'm getting the same error as @nickmccurdy but with `.storybook/main.ts`

That is to ensure `.ts` files are properly transcompiled by Babel. You do not need to use it anywhere, it is handled by Babel.

**Pull request merged**